epoll/src/main/scala/fs2/io/epoll/unsafe/EpollSystem.scala

```diff
@@ -0,0 +1,255 @@
+                  go(a, after)
+                else
+                  IO.asyncCheckAttempt[Int] { cb =>
+                    IO {
+                      writeCallback = cb
+                      // check again before we suspend
+                      val now = writeReadyCounter
+                      if (now != before) {
+                        writeCallback = null
+                        Right(now)
+                      } else Left(Some(IO(this.writeCallback = null)))
+                    }
+                  }.flatMap(go(a, _))
+              }
+            case Right(b) => IO.pure(b)
+          }
+
+        IO(writeReadyCounter).flatMap(go(a, _))
+      }
+  }
+
+  final class Poller private[EpollSystem] (epfd: Int) {
+
+    private[this] val handles: HashMap[Long, PollHandle] = HashMap()
+    private[EpollSystem] def close(): Unit =
+      if (libc.jnr.close(epfd) != 0)
+        throw new IOException(strerror(errno()))
+
+    private[EpollSystem] def poll(timeout: Long): Boolean = {
+
+      val events = Memory.allocate(globalRuntime, EpollEvent.CStructSize * MaxEvents)
+      var polled = false
+
+      @tailrec
+      def processEvents(timeout: Int): Unit = {
+
+        val triggeredEvents = epoll_wait(epfd, events, MaxEvents, timeout)
+
+        if (triggeredEvents >= 0) {
+          polled = true
+
+          var i = 0
+          while (i < triggeredEvents) {
+            val epollEvent =
+              EpollEvent.apply(events, (i * EpollEvent.CStructSize).toLong, globalRuntime)
+            val handle = handles.get(epollEvent.data)
+
+            handle
+              .getOrElse(throw new IllegalStateException("Requested fd is not found"))
+              .notify(epollEvent.events.toInt)
+
+            i += 1
+          }
+        } else if (errno() != Errno.EINTR.intValue()) { // spurious wake-up by signal
+          throw new IOException(strerror(errno()))
+        }
+
+        if (triggeredEvents >= MaxEvents)
+          processEvents(0) // drain the ready list
+        else
+          ()
+      }
+
+      val timeoutMillis = if (timeout == -1) -1 else (timeout / 1000000).toInt
+      processEvents(timeoutMillis)
+
+      polled
+    }
+
+    private[EpollSystem] def needsPoll(): Boolean = !handles.isEmpty
+
+    private[EpollSystem] def register(
+        fd: Int,
+        reads: Boolean,
+        writes: Boolean,
+        handle: PollHandle,
+        cb: Either[Throwable, (PollHandle, IO[Unit])] => Unit
+    ): Unit = {
+      val events = (EPOLLET | (if (reads) EPOLLIN else 0) | (if (writes) EPOLLOUT else 0)).toInt
+      val data = (System.identityHashCode(handle).toLong << 32) | fd
+      val event = EpollEvent(events, data, globalRuntime)
+
+      val result =
+        if (epoll_ctl(epfd, EPOLL_CTL_ADD, fd, event) != 0)
+          Left(new IOException(strerror(errno())))
+        else {
+          handles.put(data, handle)
+          val remove = IO {
+            handles.remove(data)
```

Comment:
> we could set the `PollHandle` pointer directly in epoll_event.data, like native does.

I don't think it's possible to access the value of a pointer on the JVM. But maybe I didn't understand your proposal.

> In this implementation, the `Poller` has a it's own hashmap

I think a map or set will always be necessary, because we need to keep live references to the `PollHandle`s so they are not not GCed